pkg/logs/input/container/launcher.go

```diff
@@ -6,80 +6,109 @@
+// Launcher tries to select a container launcher and retry on failure
+type Launcher struct {
+	containerLaunchables []Launchable
+	activeLauncher       restart.Restartable
+	stop                 bool
+	lock                 sync.Mutex
+}
 
-		launcher, err = docker.NewLauncher(dockerReadTimeout, sources, services, pipelineProvider, registry, false, dockerCollectFromFiles, dockerForceCollectFromFile)
-		if err == nil {
-			log.Info("Docker launcher initialized")
-			return launcher
-		}
-		log.Infof("Could not setup the docker launcher: %v", err)
-	} else {
-		launcher, err = docker.NewLauncher(dockerReadTimeout, sources, services, pipelineProvider, registry, false, dockerCollectFromFiles, dockerForceCollectFromFile)
-		if err == nil {
-			log.Info("Docker launcher initialized")
-			return launcher
-		}
-		log.Infof("Could not setup the docker launcher: %v", err)
+// NewLauncher creates a new launcher
+func NewLauncher(containerLaunchers []Launchable) *Launcher {
+	return &Launcher{
+		containerLaunchables: containerLaunchers,
+	}
+}
 
-		launcher, err = kubernetes.NewLauncher(sources, services, collectAll)
-		if err == nil {
-			log.Info("Kubernetes launcher initialized")
-			return launcher
+func (l *Launcher) launch(launchable Launchable) {
+	launcher := launchable.Launcher()
+	if launcher == nil {
+		launcher = NewNoopLauncher()
+	}
+	l.activeLauncher = launcher
+	l.activeLauncher.Start()
+}
+
+func (l *Launcher) shouldRetry() (bool, time.Duration) {
+	var retryer *retry.Retrier
+	for _, launchable := range l.containerLaunchables {
+		ok, rt := launchable.IsAvailable()
+		if ok {
+			l.launch(launchable)
+			return false, 0
+		}
+		// Hold on to the retrier with the longest interval
+		if retryer == nil || (rt != nil && retryer.NextRetry().Before(rt.NextRetry())) {
+			retryer = rt
 		}
-		log.Infof("Could not setup the kubernetes launcher: %v", err)
 	}
+	if retryer == nil {
+		log.Info("Nothing to retry - stopping")
+		return false, 0
+	}
+	nextRetry := time.Until(retryer.NextRetry())
+	log.Infof("Could not find an available a container launcher - will try again in %s", nextRetry.Truncate(time.Second))
+	return true, nextRetry
+
+}
 
-	launcher, err = docker.NewLauncher(dockerReadTimeout, sources, services, pipelineProvider, registry, true, dockerCollectFromFiles, dockerForceCollectFromFile)
-	if err != nil {
-		log.Warnf("Could not setup the docker launcher: %v. Will not be able to collect container logs", err)
-		return NewNoopLauncher()
+// Start starts the launcher
+func (l *Launcher) Start() {
+	// If we are restarting, start up the active launcher since we already picked one from a previous run
+	l.lock.Lock()
+	if l.activeLauncher != nil {
+		l.stop = true
+		l.activeLauncher.Start()
+		l.lock.Unlock()
+		return
 	}
+	l.stop = false
+	timer := time.NewTimer(0)
+	l.lock.Unlock()
 
-	log.Infof("Container logs won't be collected unless a docker daemon is eventually started")
+	// Try to select a launcher
+	go func() {
+		for {
```

Comment:
is the goal to retry ad-infinitum, should we ever give up?

Reply:
Goal is to try forever (so long as we are in a container environment). 
We actually already retry the docker launcher forever today - but it happens at the AD level. We just spawn a lazy docker launcher that can receive sources. Part of my goal with this change was to make that retry much more obvious and generalized. 

We may or may not want to make the log lines `debug` instead of `info` though.

Reply:
I think `info` is fine, it's a bad enough thing if we can't find a launcher it's got to be reported. We need to address the issue in general about verbosity with some work around the logger, but that's a different discussion altogether.